source/git-api.js

```diff
@@ -31,87 +30,125 @@ exports.registerApi = (env) => {
+        });
+        socket.watcher = watcher;
       });
     });
   }
 
-  const watchPath = (socket, subfolderPath, options) => {
-    const pathToWatch = path.join(socket.watcherPath, subfolderPath);
+  let watcherId = 1;
+  class RepoWatcher extends EventEmitter {
+    constructor() {
+      super();
+      this.watcherId = watcherId++;
+      this.watchers = [];
+    }
+    async watchItem(name, item, options = {}) {
+      if ((await fs.access(item).catch(() => false)) === false) {
+        winston.debug(`[${this.watcherId}] path does not exist`, item);
+        return;
+      }
+      const watcher = watch(item, options);
+      watcher.on('change', (_event, changedPath) => {
+        winston.silly(`[${this.watcherId}] ${name}`, changedPath);
+        this.emit(name, changedPath);
+      });
+      this.watchers.push(watcher);
+    }
+    addWorkdir(item, options) {
+      return this.watchItem('workdir', item, options);
+    }
+    addGit(item, options) {
+      return this.watchItem('git', item, options);
+    }
+    close() {
+      this.watchers.forEach((w) => w.close());
+    }
+  }
+
+  const readIgnore = async (pathToWatch) => {
+    winston.debug(`Parsing .gitignore for ${pathToWatch}`);
+    const out = ignore();
+    // For Windows/Mac that can't skip
+    out.add('.git/');
+    const ignoreContent = await fs.readFile(path.join(pathToWatch, '.gitignore')).catch(() => null);
+    if (ignoreContent) out.add(ignoreContent.toString());
+    return out;
+  };
+
+  // TODO move to nodegit
+  const watchRepo = async (pathToWatch, ignore) => {
     winston.info(`Start watching ${pathToWatch}`);
-    return fs
-      .access(pathToWatch)
-      .catch(() => {
-        // Sometimes necessary folders, '.../.git/refs/heads' and etc, are not created on git init
-        winston.debug(`intended folder to watch doesn't exists, creating: ${pathToWatch}`);
-        return mkdirp(pathToWatch);
-      })
-      .then(() => {
-        const watcher = watch(pathToWatch, options || {});
-        watcher.on('change', (event, filename) => {
-          if (!filename) return;
-          const filePath = path.join(subfolderPath, filename);
-          winston.debug(`File change: ${filePath}`);
-          if (isFileWatched(filePath, socket.ignore)) {
-            winston.info(`${filePath} triggered refresh for ${socket.watcherPath}`);
-            emitGitDirectoryChanged(socket.watcherPath);
-            emitWorkingTreeChanged(socket.watcherPath);
+    const watcher = new RepoWatcher();
+    let repoPath = path.join(pathToWatch, '.git');
+    let refsPath;
+    if ((await fs.access(repoPath).catch(() => false)) === undefined) {
+      // Looks like a repo, let's watch workdir
+      let gitIgnore = await readIgnore(pathToWatch);
+      await watcher.addWorkdir(pathToWatch, {
+        recursive: true,
+        filter: (changedPath, skip) => {
+          const filePath = path.relative(pathToWatch, changedPath);
```

Comment:
On windows it can happen that the `changedPath` is equal to `pathToWatch` which results in an empty string when `path.relative` is called. Otherwise ungit crashes when `gitIgnore.ignores(filePath)` gets called as it does not allow empty strings to be pass in.

```suggestion
          const filePath = path.relative(pathToWatch, changedPath);
          if (!filePath) return false;
```